### PR TITLE
add lock

### PR DESCRIPTION
**Pull request by Qi77Qi**
Ticket: https://broadinstitute.atlassian.net/browse/GAWB-3859

* changed docker image for test container

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/util/DistributedLockSpec.scala [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  "withLock" should "eventually get a lock with max retry" in {
+    val lockPath = genLockPath.sample.get.copy(expiresIn = 7 seconds) //fix expiresIn so that we won't be waiting for too long in the unit test
+    val res = lockResource.use { lock =>
+      for {
+        current <- timer.clock.realTime(TimeUnit.MILLISECONDS)
+        _ <- lock.acquireLock(lockPath)
+        _ <- IO.sleep(2 seconds)
+        aquireTime <- lock.withLock(lockPath).use { _ =>
```
Comment (marctalbott) [inline automation/src/test/scala/org/broadinstitute/dsde/test/util/DistributedLockSpec.scala]: spelling: acquireTime

File: project/Dependencies.scala [hunk 2]
```diff
@@ -5,12 +5,12 @@ object Dependencies {
...
-  val scalaTestV    = "3.0.1"
+  val scalaTestV    = "3.0.5"
   val scalaCheckV    = "1.14.0"
 
-  val workbenchUtilV   = "0.3-2771e2d"
-  val workbenchModelV  = "0.12-a19203d"
-  val workbenchGoogleV = "0.16-0245949"
+  val workbenchUtilV   = "0.5-36b9cda"
+  val workbenchModelV  = "0.13-7e86fba"
+  val workbenchGoogleV = "0.18-3e90c9e-SNAP"
```
Comment (marctalbott) [inline project/Dependencies.scala]: Just a reminder to update this before merging!

**Review by marctalbott — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 3]
```diff
@@ -262,13 +264,10 @@ class GoogleExtensions(directoryDAO: DirectoryDAO, val accessPolicyDAO: AccessPo
...
 
-    createPet.unsafeToFuture() recoverWith {
-      case gjre: GoogleJsonResponseException if gjre.getDetails.getCode == StatusCodes.Conflict.intValue =>
-        googleIamDAO.findServiceAccount(project, petSaName).map { saOpt =>
-          val sa = saOpt.getOrElse(throw new WorkbenchException(s"Could not create pet service account for user [${user.email}]"))
-          PetServiceAccount(PetServiceAccountId(user.id, project), sa)
-        }
-    }
+    // googleServicesConfig.resourceNamePrefix is an environment specific variable passed in https://github.com/broadinstitute/firecloud-develop/blob/fade9286ff0aec8449121ed201ebc44c8a4d57dd/run-context/fiab/configs/sam/docker-compose.yaml.ctmpl#L24
+    // Use resourceNamePrefix to avoid collision between different fiab environments (we share same firestore for fiabs)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala]: is this still the right place for this comment?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 4]
```diff
@@ -262,13 +264,10 @@ class GoogleExtensions(directoryDAO: DirectoryDAO, val accessPolicyDAO: AccessPo
...
+    val lock = LockPath(CollectionName(s"${project.value}-createPet"), Document(user.id.value), 30 seconds)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:315]: are there any limits on the length of collection or document names?

File: automation/src/test/scala/org/broadinstitute/dsde/test/SamConfig.scala [hunk 5]
```diff
@@ -1,10 +1,11 @@
-package org.broadinstitute.dsde.test
+package org.broadinstitute.dsde.workbench.test
 
 import org.broadinstitute.dsde.workbench.config.CommonConfig
 
 object SamConfig extends CommonConfig {
   // from common: qaEmail, pathToQAPem
   object GCS extends CommonGCS {
     val appsDomain = gcsConfig.getString("appsDomain")
+    val pathToSamTestFirestoreAccountPath = gcsConfig.getString("firestoreAccountPath")
```
Comment (dvoet) [inline automation/src/test/scala/org/broadinstitute/dsde/test/SamConfig.scala:9]: where is this populated? In our unit tests do we use a real account talking to real firestore? Is there any way to mock that out? Is there a local firestore emulator?

https://firebase.google.com/docs/firestore/security/test-rules-emulator#install_the_emulator

**Review by dvoet — CHANGES_REQUESTED** [review]
I think the code is fine but I would like to find a way to run our tests without a real firestore. Perhaps we can use
https://firebase.google.com/docs/firestore/security/test-rules-emulator#install_the_emulator

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala (hunk 4 above)
Comment (Qi77Qi) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:315]: seems `1,500 bytes` is the limit according to https://firebase.google.com/docs/firestore/quotas

File: automation/src/test/scala/org/broadinstitute/dsde/test/SamConfig.scala (hunk 5 above)
Comment (Qi77Qi) [reply automation/src/test/scala/org/broadinstitute/dsde/test/SamConfig.scala:9]: it's populated https://github.com/broadinstitute/firecloud-automated-testing/blob/master/configs/sam/application.conf.ctmpl#L47.

I wasn't aware there's emulator. I'll check it out. But I guess that can be in another PR?

**Review by dvoet — APPROVED** [review]

**Pull request merged**